sklearn/utils/tests/test_validation.py

```diff
@@ -1675,9 +1675,9 @@ def test_get_feature_names_invalid_dtypes(names, dtypes):
     X = pd.DataFrame([[1, 2], [4, 5], [5, 6]], columns=names)
 
     msg = re.escape(
-        "Feature names only support names that are all strings. Got feature names with"
-        f" dtypes: {dtypes}. Please convert to a common type, for example using"
-        " X.columns = X.columns.astype(str)"
+        "Feature names only support column names that are all strings, but got dtypes:"
```

Comment:
nit:
```suggestion
        "Feature names only support column names that are strings, but got dtypes:"
```
The "all" is unnecessary and I find it confuses me more/makes for tricky reading/doesn't make things easier to understand.

Reply:
I think w/o the `all` here, it could mean that only string feature names are supported and the rest can be ignored. The two don't mean the same thing to me, and the latter doesn't seem correct.

Reply:
The way I understood the feature is that if you specify a column name, it has to be a string. No other types are allowed. Is that right? If not ignore the below.

For me "column names that are all strings"  is like "food made from all organic ingredients". The "all" is a statement about the ingredients, which in this case are exclusively organic. As a single column name can't consist of two different types it seems weird to say "all strings". "All column names have to be strings" is a statement about the column names, they all have to be strings.

Reply:
there's a logical fallacy there in your example. feature names are separate things, but food is one thing, therefore your example would be more like "food made from only organic ingredients", or a better example: "we can allow a group of people who are all above 18", and to me it makes senes.

Reply:
From reading the code I think the answer to my question is: no.

This means that after considering only the exception's message and the few lines of code around it I didn't understand what the problem was. Hence I don't have a good idea how to succinctly explain to a user what just happened and what they need to do to fix it.

Reply:
Maybe we could write:

> Feature names are only supported if all input features have string names, but your input has {types} as feature name / column name types. If you want feature names to be stored and validated, you must convert them all to strings. If you'd like to silence this warning, you can remove feature / column names from your input data, or convert them all to a non-string data type.

Reply:
I think this ☝️ is an improvement over what we currently have

Reply:
I directly pushed the change to use Adrin's suggestion because it looks uncontroversially more clear and I think it's still the middle of the night for Thomas :)

Reply:
I removed the last sentence about silencing the warning since it's not a warning actually, good catch from https://github.com/scikit-learn/scikit-learn/pull/25018#discussion_r1032272884

Reply:
I think we should tell people how to disable it though. They don't have to pass feature names, this makes it look like they have to.

Reply:
I see, I put it back but without mentioning the warning. Is it fine now ?